### PR TITLE
Cross-block persistent support in codegen

### PR DESCRIPTION
**Pull request by csarofeen**
Adds:
Cooperative launch support
Grid broadcast
Grid reductions in loops
Grid reduction -> grid broadcast patterns

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 1]
```diff
@@ -813,19 +833,35 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
...
+          launch_params.bdimy(),
+          launch_params.bdimz(),
+          launch_params.smem(),
+          stream,
+          kernel_arguments.getBuffer(),
+          nullptr));
+    } else {
+      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchCooperativeKernel");
+      AT_CUDA_DRIVER_CHECK(
+          at::globalContext().getNVRTC().cuLaunchCooperativeKernel(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/executor.cpp:856]: nvfuser is also enabled for rocm. cuLaunchCooperativeKernel is only added for cuda, so we need to guard the symbol here with `#ifndef __HIP_PLATFORM_HCC__` and assert on the else block.

File: torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu [hunk 2]
```diff
@@ -188,9 +188,10 @@ template <
     bool X_THREAD,
     bool Y_THREAD,
     bool Z_THREAD,
+    bool PERSISTENT_REDUCTION,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu:198]: Can you please add a comment on this new template parameter?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 3]
```diff
@@ -999,6 +1010,48 @@ class CudaKernelGenerator : private kir::IrVisitor {
...
+      const bool parallel_bcast = par_domains.find(pt) != par_domains.end() &&
+          par_domains.at(pt)->isBroadcast();
+      if (pt != kParallelTypeThreads[0]) {
+        flags_str << ", ";
+      }
+      flags_str << (parallel_bcast ? "true" : "false");
+    }
+
+    // Since block-level reduction is already done, those dimensions
+    // with tidx/y/z being true do not participate in the grid reduction.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/codegen.cpp]: Copied from gridReduction?

File: torch/csrc/jit/codegen/cuda/runtime/welford.cu [hunk 4]
```diff
@@ -393,9 +393,10 @@ template <
     bool X_THREAD,
     bool Y_THREAD,
     bool Z_THREAD,
+    bool PERSISTENT_REDUCTION,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/welford.cu:273]: Isn't this function also supposed to use this parameter?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 5]
```diff
@@ -1019,13 +1072,16 @@ class CudaKernelGenerator : private kir::IrVisitor {
...
+
     const std::string flags_str =
         generateGridReduceTemplateFlags(wop, node->threadPredicate());
 
     // Since block-level reduction is already done, those dimensions
     // with tidx/y/z being true do not participate in the grid reduction.
-    indent() << kir::GridWelford::getPredicateFlagName(out->view()) << " = "
-             << "welford::gridWelford<" << flags_str << ">(\n";
+    indent() << "welford::gridWelford<" << flags_str << ", "
+             << (persistent_sync ? "true" : "false") << ">(\n";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:1085]: The persistent flag doesn't seem to be used in `gridWelford`. Is it intended?

File: torch/csrc/jit/codegen/cuda/runtime/grid_broadcast.cu [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    bool Y_THREAD,
+    bool Z_THREAD,
+    typename T>
+__device__ void broadcast(
+    T& out,
+    const T& inp_val,
+    volatile T* work_buf,
+    Tensor<int64_t, 1> sync_flags,
+    bool read_write_pred) {
+  // Number of values to reduce in the grid dimensions
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/grid_broadcast.cu]: This comment needs to be adjusted as this isn't a reduction.

File: torch/csrc/jit/codegen/cuda/runtime/grid_broadcast.cu [hunk 7]
```diff
@@ -0,0 +1,55 @@
+namespace grid_broadcast {
+template <
+    bool X_BLOCK,
+    bool Y_BLOCK,
+    bool Z_BLOCK,
+    bool X_THREAD,
+    bool Y_THREAD,
+    bool Z_THREAD,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/grid_broadcast.cu:22]: Can you please add comments here?

File: test/cpp/jit/test_gpu.cpp [hunk 8]
```diff
@@ -6747,6 +6747,88 @@ TEST(NVFuserTest, FusionGridReduction8_CUDA) {
   testValidate(&fusion, out, {input}, {aten_output}, __LINE__, __FILE__);
 }
 
+TEST(NVFuserTest, FusionGridReduction9_CUDA) {
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: I think it's also good to have tests for grid broadcast.

**Review by naoyam — COMMENTED** [review]
Added some comments. Mostly minor cleanups. I'd approve once the ROCM compatibility is fixed.

File: test/cpp/jit/test_gpu.cpp (hunk 8 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: Grid persistence tests are testing grid broadcast, since we need a reduction before the broadcast to trigger it.

Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: Do you want more grid broadcast tests?

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp]: I was thinking of a bare minimum example, but `FusionGridPersistence` seems to be the simplest case. Looks sufficient to me.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for addressing the review comments.

**Review by jjsjann123 — APPROVED** [review]

**Pull request merged**